### PR TITLE
feat(speechlm2): add padding-free packed speech training

### PR DESCRIPTION
**Pull request by pzelasko**
# feat(speechlm2): add padding-free packed speech training

> **Stack 3/7:** [1/7 Encoder foundations](https://github.com/NVIDIA-NeMo/Speech/pull/16192) → [2/7 Indexed/resumable data](https://github.com/NVIDIA-NeMo/Speech/pull/16193) → **Packed speech training** → [4/7 Portable export + target-only vLLM](https://github.com/NVIDIA-NeMo/Speech/pull/16195) → [5/7 MTP serving](https://github.com/NVIDIA-NeMo/Speech/pull/16196) → [6/7 DFlash serving](https://github.com/NVIDIA-NeMo/Speech/pull/16239) → [7/7 CCE dependency](https://github.com/NVIDIA-NeMo/Speech/pull/16241)

Base: `main`

Head: `codex/pr16158-rewrite/03-packed-speech-training`

Tip: `919c31b5bf65f83e27e57cbe1f853cf06a8090e4`

## Summary

- Add native THD Transformer/PEE execution and packed activation, waveform, feature, and SpecAugment paths.
- Add padding-free SALM batches with CP/FSDP/FP8 alignment and activation checkpointing.
- Add PEE execution controls, fused loss, and the FSDP-shared MTP training loss head.
- Add optimizer-resume, garbage-collection, and OOMptimizer runtime hardening.
- Keep the obsolete GGEMM/MoE implementations, benchmarks, imports, and tests removed throughout this layer.

## Reviewer guide

The review question is: does SALM consume the encoder and data foundations correctly under packed, distributed, resumable training?

Index-format and routing implementations remain in PR 2. Checkpoint export and all vLLM behavior remain outside this PR.

## Attribution

The replay preserves original authors, dates, source-commit references, and author-matching DCO sign-offs.

## Validation

- `git diff --check`: passed.
- Compilation: passed.
- Focused stack suite: 403 passed, 20 skipped.
- Nithin review coverage: 207 packed ASR/SpeechLM2 tests passed; the packed Transformer GPU suite passed 53/53 on GB300.
- Packed perception regression coverage passes on CPU (9 passed, 2 CUDA-only skipped) and under the GPU CI collection environment on GB300 (11 passed).
- Removed the redundant Docker-only `cut-cross-entropy` metadata/import assertion; the optional code paths remain usable without the dependency, while its declaration and lock metadata are intentionally deferred to PR 7.
- The FlashAttention device/provider cache regression test passes, including repeated dispatch on the same CUDA device.
- All newly introduced packed-attention entry points and private helpers now document their contracts and behavior.
- Rebuilt top-of-stack vLLM/export regression suite: 188 passed, 3 skipped.
- Branch-scoped and complete-stack pre-commit: passed.
- `uv lock --check --offline`: passed.
- Range-diff confirms unchanged commits are patch-identical; the only deviations are the intentional obsolete-component cleanup and its downstream references.
- Documentation is unchanged because this is a private helper extraction and runtime dispatch optimization with no public API or configuration change.

### CODE REVIEW

File: tests/collections/asr/test_parallel_expert_encoder_two_branch.py [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+    ParallelExpertEncoderPT,
+    _clone_config,
+    _default_dtype,
+    _disable_dist_feature_sync,
+)
+from nemo.collections.asr.modules.transformer_encoder import TransformerEncoder
+from nemo.collections.asr.parts.packed_sequence import pack_encoder_output, unpack_encoder_output
+from nemo.collections.asr.parts.preprocessing.features import normalize_batch, normalize_packed_batch
+
+_PEE = getattr(ParallelExpertEncoder, "__wrapped__", ParallelExpertEncoder)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/asr/test_parallel_expert_encoder_two_branch.py]: Resolved in a0eb08d6bc7d427ebf710f282c7c915d0819da76 by removing the unused _PEE test alias.

File: nemo/collections/asr/modules/parallel_expert_encoder_ggemm.py [hunk 2]
```diff
@@ -95,18 +95,24 @@
 from tqdm import tqdm
 
 from nemo.collections.asr.modules.ggemm_transformer_encoder import GGEMMTransformerEncoder
-from nemo.collections.asr.parts.preprocessing.features import normalize_batch
+from nemo.collections.asr.modules.parallel_expert_encoder_resolver import classify_parallel_expert_encoder_config
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/asr/modules/parallel_expert_encoder_ggemm.py]: Resolved in a0eb08d6bc7d427ebf710f282c7c915d0819da76: pure PEE schema classification and bundle reading now live in a dependency-neutral schema module used by both the GGEMM implementation and concrete-class resolver, removing the cycle.

File: tests/collections/asr/test_parallel_expert_encoder_two_branch.py [hunk 3]
```diff
@@ -0,0 +1,688 @@
...
+import nemo.collections.asr.modules.parallel_expert_encoder as pee_module
+from nemo.collections.asr.models import SortformerEncLabelModel
+from nemo.collections.asr.modules.conformer_encoder import ConformerEncoder
+from nemo.collections.asr.modules.parallel_expert_encoder_two_branch import (
+    ParallelExpertEncoder,
+    ParallelExpertEncoderPT,
+    _clone_config,
+    _default_dtype,
+    _disable_dist_feature_sync,
+)
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply tests/collections/asr/test_parallel_expert_encoder_two_branch.py]: Resolved in a0eb08d6bc7d427ebf710f282c7c915d0819da76 by removing the unused ParallelExpertEncoderPT test import.

File: nemo/collections/speechlm2/parts/pretrained.py [hunk 4]
```diff
@@ -325,12 +328,96 @@
...
+
+def setup_independent_speaker_encoder(model: torch.nn.Module):
+    """Add a standalone speaker Transformer beside the pretrained ASR encoder.
+
+    ``model.speaker_encoder.path`` points at a rendered artifact directory with
+    ``model_config.yaml`` and ``model.safetensors``. The two encoders execute
+    independently inside :class:`IndependentDualEncoder`; their same-rate states
+    are concatenated before the existing perception-to-LLM projection.
+    """
+    from nemo.collections.speechlm2.modules.perception import IdentityConnector, IndependentDualEncoder
```
[the remark this answers is not in the document]
Comment (pzelasko) [reply nemo/collections/speechlm2/parts/pretrained.py:318]: Resolved in a0eb08d6bc7d427ebf710f282c7c915d0819da76: generic NeMo model-loading helpers now live in a dependency-neutral model_loading module, so perception no longer imports the setup module that imports perception.

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 5]
```diff
@@ -463,6 +475,151 @@ def forward_streaming(self, kv_normed, num_cur, block_mask, pos_emb):
...
+        *,
+        lengths,
+        cu_seqlens,
+        max_seqlen,
+        pos_emb,
+        padded_length,
+        causal,
+        sequence_offsets,
+    ):
+        use_flash = not self._uses_rel_pos and _can_use_flash_attention_varlen(q)
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder.py]: do we need to check this for each forward step?

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 6]
```diff
@@ -1380,3 +1736,141 @@ def mask_mod(b, h, q_idx, kv_idx):
...
+
+
+def _apply_packed_rope(rope, q, k, position_ids):
+    cos = rope.cos.index_select(0, position_ids).unsqueeze(1).to(q.dtype)
+    sin = rope.sin.index_select(0, position_ids).unsqueeze(1).to(q.dtype)
+    return rope._apply_rotary(q, cos, sin), rope._apply_rotary(k, cos.to(k.dtype), sin.to(k.dtype))
+
+
+def _packed_flex_attention_reference(
+    attn,
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder.py]: Can we move all these functions to seperate utils file

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 5 above)
Comment (pzelasko) [reply nemo/collections/asr/modules/transformer_encoder.py]: Addressed in 0a46d8328f. The expensive device capability and FlashAttention provider resolution are now cached once per concrete CUDA device, while the per-forward checks remain limited to tensor-dependent state that can legitimately change (device, autocast dtype, and empty packed input). The static relative-position/head-dimension eligibility is cached on module construction. I also added a regression test proving that repeated dispatch on the same CUDA device probes capability/provider only once.

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 6 above)
Comment (pzelasko) [reply nemo/collections/asr/modules/transformer_encoder.py]: Addressed in 0a46d8328f. I moved all seven packed-attention helper functions into the private sibling module `transformer_encoder_utils.py` and kept a qualified, one-way import from `transformer_encoder.py`, with no public re-export or circular dependency. The packed ASR/SpeechLM2 suite passes after the extraction.

File: nemo/collections/asr/modules/transformer_encoder_utils.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+                False,
+                scale=softmax_scale,
+            )[0]
+
+        torch_flash_attention_varlen._sequence_packed_provider = "aten"
+        return torch_flash_attention_varlen
+    return flash_attn_varlen_func
+
+
+def _forward_sequence_packed_layer(layer, x, **kwargs):
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder_utils.py:177]: are we adding these utils to fastconformer as well?

Another way of asking: Are we supporting packed sequence feature to FastConformer based encoders as well?

File: nemo/collections/asr/modules/transformer_encoder_utils.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder_utils.py:2]: do we need SPDX in header?

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 9]
```diff
@@ -458,11 +471,157 @@ def forward_streaming(self, kv_normed, num_cur, block_mask, pos_emb):
         if k_mod is not None:
             k = k_mod
 
-        attn_fn = flex_attention_compiled if q.is_cuda else flex_attention
+        attn_fn = _transformer_utils._get_flex_attention(q)
         out = attn_fn(q, k, v, block_mask=block_mask, score_mod=score_mod)
         out = out.transpose(1, 2).contiguous().view(B, num_cur, self.d_model)
         return self.out_proj(out)
 
+    def forward_sequence_packed(
```
Comment (nithinraok) [inline nemo/collections/asr/modules/transformer_encoder.py:479]: many doesn;t have doc strings, add wherever required.

File: nemo/collections/asr/modules/transformer_encoder_utils.py (hunk 8 above)
Comment (pzelasko) [reply nemo/collections/asr/modules/transformer_encoder_utils.py:2]: Yes, new rule.

File: nemo/collections/asr/modules/transformer_encoder_utils.py (hunk 7 above)
Comment (pzelasko) [reply nemo/collections/asr/modules/transformer_encoder_utils.py:177]: Not currently, but nothing stops us from backporting later. We should IMO.

File: nemo/collections/asr/modules/transformer_encoder.py (hunk 9 above)
Comment (pzelasko) [reply nemo/collections/asr/modules/transformer_encoder.py:479]: Addressed in 25c5e7b39d. I documented all eight newly introduced packed private methods in `transformer_encoder.py`, expanded both public `forward_sequence_packed()` entry points with their argument and return contracts, and documented the extracted helper module plus every top-level helper and the nested ATen adapter. I kept the change documentation-only and limited to the newly introduced packed-attention code. The affected ASR API page now passes a warning-as-error Sphinx build, and the packed Transformer suite remains green.

File: docker/Dockerfile [hunk 10]
```diff
@@ -225,6 +225,15 @@ uv sync \
     --group test
 nemo-install-cuda-python
 
+python - <<'PY'
+from importlib.metadata import version
+
+import cut_cross_entropy
```
Comment (nithinraok) [inline docker/Dockerfile]: why is this required?

**Review by nithinraok — DISMISSED** [review]
Thanks. LGTM

File: docker/Dockerfile (hunk 10 above)
Comment (pzelasko) [reply docker/Dockerfile]: Big memory savings by avoiding logit materialization in CE loss.

Comment (nithinraok) [reply docker/Dockerfile]: Thanks, that explains why we use it. My question was about the version check in the Dockerfile: can we pin this in pyproject/uv.lock instead?

Comment (pzelasko) [reply docker/Dockerfile]: Agreed — removed in 3d8530f293. Both compiled extras already pin cut-cross-entropy 25.3.2, pyproject.toml pins its exact source revision, and uv.lock plus uv sync --locked enforce that installation. The deleted import/version assertion duplicated those guarantees and did not exercise the fused GPU kernel.

Comment (nithinraok) [reply docker/Dockerfile]: can we do same for `automodel` version check.

**Review by nithinraok — DISMISSED** [review]

**Review by chtruong814 — CHANGES_REQUESTED** [review]
We'll need to review the inclusion of cut-cross-entropy.

**Review by chtruong814 — DISMISSED** [review]

**Review by nithinraok — DISMISSED** [review]
